src/Voron/Data/Tables/TableValueCompressor.cs

```diff
@@ -0,0 +1,401 @@
+
+            using var compressionDictionary = new ZstdLib.CompressionDictionary(newId, dictionaryBuffer.Ptr, dictionaryBufferSpan.Length, 3);
+
+            if (ShouldReplaceDictionary(tx, compressionDictionary) == false)
+            {
+                return;
+            }
+
+            table.CurrentCompressionDictionaryId = newId;
+            compressionDictionary.ExpectedCompressionRatio = GetCompressionRatio(CompressedBuffer.Length, RawBuffer.Length);
+
+            var rev = Bits.SwapBytes(newId);
+            using var _____ = Slice.External(tx.Allocator, (byte*)&rev, sizeof(int), out var slice);
+            using var ____ = dictionariesTree.DirectAdd(slice, sizeof(CompressionDictionaryInfo) + dictionaryBufferSpan.Length, out var dest);
+            *((CompressionDictionaryInfo*)dest) = new CompressionDictionaryInfo
+            {
+                ExpectedCompressionRatio = compressionDictionary.ExpectedCompressionRatio
+            };
+            Memory.Copy(dest + sizeof(CompressionDictionaryInfo), dictionaryBuffer.Ptr, dictionaryBufferSpan.Length);
+
+            tx.LowLevelTransaction.OnDispose += RecreateRecoveryDictionaries;
+        }
+
+        public static readonly byte[] EncryptionContext = Encoding.UTF8.GetBytes("Compress");
+
+        private static void RecreateRecoveryDictionaries(IPagerLevelTransactionState obj)
+        {
+            if (!(obj is LowLevelTransaction llt) || llt.Committed == false)
+                return; // we can't write on non committed transactions
+
+            if (obj.Environment.Options is StorageEnvironmentOptions.PureMemoryStorageEnvironmentOptions)
+                return; // no need for in mem mode
+
+            lock (obj.Environment.CompressionDictionariesHolder)
+            {
+                using var tx = obj.Environment.ReadTransaction();
+
+                var dictionaries = tx.ReadTree(TableSchema.DictionariesSlice);
+
+                if (dictionaries == null)
+                    return; // should never happen
+
+                int nonceSize = (int)Sodium.crypto_stream_xchacha20_noncebytes();
+                var subKeyLen = (int)Sodium.crypto_aead_xchacha20poly1305_ietf_keybytes();
+                int macSize = (int)Sodium.crypto_aead_xchacha20poly1305_ietf_abytes();
+                var nonceBuffer = stackalloc byte[nonceSize];
+                var macBuffer = stackalloc byte[macSize];
+                var subKey = stackalloc byte[subKeyLen];
+
+                // for reliability's sake, we keep two copies of the compression
+                // dictionaries
+                for (int i = 0; i < 2; i++)
+                {
+                    var newPath = obj.Environment.Options.BasePath
+                        .Combine(path: $"Compression{(i == 0 ? "A" : "B")}.Recovery")
+                        .FullPath;
+
+                    using var fs = File.Open(newPath, FileMode.OpenOrCreate, FileAccess.ReadWrite, FileShare.Read);
```

Comment:
Don't we need to call `fs.Flush(flushToDisk: true);` at the end?

Reply:
I intentionally didn't do that. 
We have two separate files here, and I want to avoid flushing the entire fs cache
That can kill our perf

Reply:
1. But if non of the files is flushed then how can we be guaranteed that at least one have everything synced? 
2. This is called once per write transaction, right?

Reply:
This is called per value saved.

Reply:
Yes, sorry, once per write tx that created new dictionary
And we don't flush because:
* We have the data saved in the data file
* We re-create it on new dictionarys